plugins/auth-backend/src/service/OidcService.ts

```diff
@@ -89,4 +108,404 @@ export class OidcService {
+
+    const allowedRedirectUriPatterns = this.config.getOptionalStringArray(
+      'auth.experimentalDynamicClientRegistration.allowedRedirectUriPatterns',
+    );
+
+    if (allowedRedirectUriPatterns) {
+      for (const redirectUri of opts.redirectUris ?? []) {
+        if (
+          !allowedRedirectUriPatterns.some(pattern =>
+            new RegExp(pattern).test(redirectUri),
+          )
+        ) {
+          throw new InputError('Invalid redirect_uri');
+        }
+      }
+    }
+
+    return await this.oidc.createClient({
+      clientId: generatedClientId,
+      clientName: opts.clientName,
+      clientSecret: generatedClientSecret,
+      redirectUris: opts.redirectUris ?? [],
+      responseTypes: opts.responseTypes ?? ['code'],
+      grantTypes: opts.grantTypes ?? ['authorization_code'],
+      scope: opts.scope,
+    });
+  }
+
+  public async createAuthorizationSession(opts: {
+    clientId: string;
+    redirectUri: string;
+    responseType: string;
+    scope?: string;
+    state?: string;
+    nonce?: string;
+    codeChallenge?: string;
+    codeChallengeMethod?: string;
+  }) {
+    const {
+      clientId,
+      redirectUri,
+      responseType,
+      scope,
+      state,
+      nonce,
+      codeChallenge,
+      codeChallengeMethod,
+    } = opts;
+
+    if (responseType !== 'code') {
+      throw new InputError('Only authorization code flow is supported');
+    }
+
+    const client = await this.oidc.getClient({ clientId });
+    if (!client) {
+      throw new InputError('Invalid client_id');
+    }
+
+    if (!client.redirectUris.includes(redirectUri)) {
+      throw new InputError('Invalid redirect_uri');
+    }
+
+    if (codeChallenge) {
+      if (
+        !codeChallengeMethod ||
+        !['S256', 'plain'].includes(codeChallengeMethod)
+      ) {
+        throw new InputError('Invalid code_challenge_method');
+      }
+    }
+
+    const sessionId = crypto.randomUUID();
+    const sessionExpiresAt = DateTime.now().plus({ hours: 1 }).toJSDate();
+
+    await this.oidc.createAuthorizationSession({
+      id: sessionId,
+      clientId,
+      redirectUri,
+      responseType,
+      scope,
+      state,
+      codeChallenge,
+      codeChallengeMethod,
+      nonce,
+      expiresAt: sessionExpiresAt,
+    });
+
+    return {
+      id: sessionId,
+      clientName: client.clientName,
+      scope,
+      redirectUri,
+    };
+  }
+
+  public async approveAuthorizationSession(opts: {
+    sessionId: string;
+    userEntityRef: string;
+  }) {
+    const { sessionId, userEntityRef } = opts;
+
+    const session = await this.oidc.getAuthorizationSession({
+      id: sessionId,
+    });
+
+    if (!session) {
+      throw new NotFoundError('Invalid authorization session');
+    }
+
+    if (DateTime.fromJSDate(session.expiresAt) < DateTime.now()) {
+      throw new InputError('Authorization session expired');
+    }
+
+    if (session.status !== 'pending') {
```

Comment:
arguably these types of invariant checks should be all the way inside `oidc.updateAuthorizationSession`, using database query semantics to ensure it's checked atomically (`WHERE status = 'pending'` etc) and handling all failed such updates under the "Authorization session not found or expired" umbrella. Same in the other methods below